src/WebSockets/SubscriptionProtocolHandler.cs

```diff
@@ -79,11 +81,47 @@ public Task HandleConnectionClosed(OperationMessageContext context)
 
         protected async Task HandleStartAsync(OperationMessageContext context)
         {
-            var query = context.Op.Payload.ToObject<GraphQLQuery>();
-            var result = await SubscribeAsync(query, context).ConfigureAwait(false);
+            var payload = context.Op.Payload;
+            var query = payload is GraphQLQuery ? payload : context.Op.Payload.ToObject<GraphQLQuery>();
+            var options = context.Connection.Options;
+            var exOptions = new ExecutionOptions
+            {
+                Schema = _schema,
+                OperationName = query.OperationName,
+                Inputs = query.GetInputs(),
+                Query = query.Query,
+                ExposeExceptions = options?.ExposeExceptions ?? false,
+                ValidationRules = options?.ValidationRules,
+                UserContext = options?.BuildUserContext?.Invoke(context)
+            };
+
+            var isSubscription = _determinator.IsSubscription(exOptions);
+
+            if (isSubscription)
```

Comment:
I would probably implement this so that even when executing mutation or query I would still handle it as an subscription and just wrap the result into `ReplaySubject` and complete it after the first (only) value. This would keep the pipeline clean and allow other future enhancements without burden of forked logic.

Reply:
I disagree that this should be handled as a subscription with only one value. I feel that the name `SubscriptionProtocolHandler.cs` is actually wrong and should be called something like `WebSocketProtocolHandler.cs`.

My reasoning here is that currently this file is handling messages which happen to be coming from a websocket. The messages just happen to be a subscription, mutation or a query. Someone could in theory go and implement a `HttpLongPoolingHandler.cs` which would also need to handle subscriptions, mutations and queries.

At a framework level, subscriptions were different enough to require their own handler independent from queries/mutations. So I feel that this protocol handler should call the appropriate handler and return responses appropriately. 

I am happy to look at how to modify this code, but unless the need for a separate handler for subscriptions is removed I feel something along these lines is appropriate.

Reply:
Maybe I've been been too deep in these lately but WebSocket protocol to me is the actual protocol that handles the transfer between the browser and server and does not have anything to do with GraphQL. 

I chose the name SubscriptionProtocolHandler because ApolloSubscriptionProtocolHandler was getting bit too long. That's what this is. It handles subscriptions according to how Apollo handles them. Being compatible with Apollo client.

Reply:
The reasoning for my agrument actually comes from that project as well. Following the protocol information provided here: https://github.com/apollographql/subscriptions-transport-ws/blob/master/PROTOCOL.md . 

Please correct me if I am miss understanding this in any way, but this protocol seems to define a mechanism of sending graphql requests over a websocket channel and doesn't specifically require the payload to be a subscription. When I read the Message flow section, specifically the Connected Phase, I tried to emulate that in the code above. 

I think that the subscription-transport-ws has also grown past the subscription requirement is really is a transport-ws library now.